# 4.5 Data Consistency Checks

## This script contains the following points:

### 0.1 Import Libraries

### 0.2 Import Data

### 0.3 Data Consistency Checks
#### 0.3.1 Mixed Data Types
#### 0.3.2 Missing Data Values
#### 0.3.3 Find Duplicate Values
#### 0.3.4 Addressing Duplicate Values
#### Data Cleaning Addition from Exercise 4.9 (descriptive statistics and loc() function)

### 0.4 Export Data

### 0.5 Task
#### Question 2
#### Question 3
#### Question 4
#### Question 5
#### Question 6
#### Question 7
#### Question 8
#### Question 9

# 0.1 Import Libraries

In [1]:
import pandas as pd
import numpy as np
import os

# 0.2 Import Data

In [2]:
# Creating path string to data folder
path = r'C:\Users\bexlu\Desktop\Career Foundry\Data Immersion\Achievement 4\Instacart Basket Analysis'

In [3]:
# Importing products_csv file using os library
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# Importing orders_wrangled file using os library
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

# 0.3 Data Consistency Checks

## 0.3.1 Mixed Data Types

In [5]:
# Create a practice dataframe
df_test = pd.DataFrame()

In [6]:
# Create a mixed data type column within df_test
df_test['mix'] = ['a','b',1,True]

In [7]:
# Visualise mix column
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Apply advanced python function that checks df for mixed type columns
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
# Change data type of mix column to string
df_test['mix'] = df_test['mix'].astype('str')

In [10]:
# Check data type of mix column
df_test['mix'].dtype

dtype('O')

## 0.3.2 Missing Data Values

In [11]:
# Find missing values in df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [12]:
# Create new subset with missing values from product_name
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [13]:
# Visualise subset
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [15]:
# Check number of rows of df_prods before create subset without missing values
# By checking rows we will be able to compare the shape of df_prods to new subset
df_prods.shape

(49693, 5)

In [16]:
# Create new df excluding missing values from the 'product name' column
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [17]:
# Check shape of df_prods_clean to compare shape to df_prods
df_prods_clean.shape

(49677, 5)

## 0.3.3 Find Duplicate Values

In [18]:
# Search for duplicated rows in df_prods_clean and create a subset that contains only these rows
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [20]:
# Visualise df_dups subset that contains duplicated rows
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

## 0.3.4 Addressing Duplicate Values

In [21]:
# Check number of rows in df so we can compare after removing duplicates
df_prods_clean.shape

(49677, 5)

In [22]:
# Create new df that excludes duplicated rows
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [23]:
# Check number of rows of new df that excludes duplicated rows
df_prods_clean_no_dups.shape

(49672, 5)

## Data Cleaning Addition from exercise 4.9

In [1]:
# finding errors in pricing column setting 

# Check data consistency
ords_prods_merge['prices'].describe()

ords_prods_merge['prices'].mean

ords_prods_merge['prices'].mean

ords_prods_merge['prices'].max()

# Explore values within prices column that are greater than 100 using loc() function
ords_prods_merge.loc[ords_prods_merge['prices']>100]

# mark outlying values as missing using np.nan function
ords_prods_merge.loc[ords_prods_merge['prices'] > 100, 'prices'] = np.nan

# check max value of prices column for comparison
ords_prods_merge['prices'].max()

# 0.4 Exporting Data

In [24]:
# Export df_prods_clean_no_dups and save to 'Prepared Data'
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

# 0.5 Task

## Question 2

In [25]:
# Print of descriptive statistics of df_prods
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

The prices column requires further investigation as the maximum price value of 99999 seems very high for a supermarket product.

## Question 3

In [26]:
# Check df_ords for mixed type data
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## Question 4

No columns were found that contained mix type data.

## Question 5

In [27]:
# Check for missing values in df_ords
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
orders_day_of_week             0
hour_order_placed              0
days_since_prior_order    206209
dtype: int64

The only column that contains missing values in df_ords is 'days_since_prior_order'. I believe these missing values exist as they represent each customers first order. I would expect that when the 'order_number' is '1', the days_since_prior_order is 'NaN'. 

In [32]:
# Find all orders where order number is 1.
df_ords.loc[df_ords['order_number']==1]

Unnamed: 0  order_id  user_id eval_set  order_number  \
0                 0   2539329        1    prior             1   
11               11   2168274        2    prior             1   
26               26   1374495        3    prior             1   
39               39   3343014        4    prior             1   
45               45   2717275        5    prior             1   
...             ...       ...      ...      ...           ...   
3420930     3420930    969311   206205    prior             1   
3420934     3420934   3189322   206206    prior             1   
3421002     3421002   2166133   206207    prior             1   
3421019     3421019   2227043   206208    prior             1   
3421069     3421069   3154581   206209    prior             1   

         orders_day_of_week  hour_order_placed  days_since_prior_order  
0                         2                  8                     NaN  
11                        2                 11                     NaN  
26                        1                 14                     NaN  
39                        6                 11                     NaN  
45                        3                 12                     NaN  
...                     ...                ...                     ...  
3420930                   4                 12                     NaN  
3420934                   3                 18                     NaN  
3421002                   6                 19                     NaN  
3421019                   1                 15                     NaN  
3421069                   3                 11                     NaN  

[206209 rows x 8 columns]

We can see that, by returning all order records where the 'order_number' is '1', we have found 206209 rows. This is the same
number of rows that contain missing data in the 'days_since_prior_order' column. 

## Question 6

As we have seen in Question 5, the data that is missing from the 'days_since_prior_order' isn't actually missing data at all, rather it represents the customers first order.
We could overwrite the 'NaN' values with '0' so that it is clear that these values are not missing, but I wouldn't necessarily do this without checking with the client first. 

## Question 7

In [39]:
# Check number of rows in df_ords
df_ords.shape

(3421083, 8)

In [40]:
# Create subset of df_ords that excludes full duplicates
df_dups = df_ords[df_ords.duplicated()]

In [41]:
# Check number of rows in subset df_dups
df_dups.shape

(0, 8)

We can see from the above output that there are no rows in the df_dups subset, this means we have no duplicate values.

## Question 8

I didn't find any duplicate values in df_ords

## Question 9

I have already exported my cleaned df_prods (see 0.4 Exporting Data).
I have not peformed any cleaning on my df_ords so have not exported it here.